### PR TITLE
Enable running without bias and update ffn instantiation

### PR DESCRIPTION
**Pull request by vchiley**
Clone of https://github.com/microsoft/tutel/pull/210 but this leaves `_num_global_experts` as a registered buffer.

This PR
- Clean up __init__ / param init for FusedExpertsNetwork
- enable FusedExpertsNetwork to run without bias

Notes:
All tensor naming is still the same. If bias=True the generated state dict will be the same. All old state dicts can still be loaded.

This only makes layer init more conventional (along with using a more conventional `reset_parameters`) eg [torch.nn.Linear layer](https://pytorch.org/docs/stable/_modules/torch/nn/modules/linear.html#Linear).
(Standard PyTorch attaches params in \_\_init__ instead of in an auxiliary `self.update(ctx)` fn. After params are attached standard PyTorch calls `reset_parameters`.)

I just moved parameter init, I didn't change how init happens. The seeding is unchanged. The parameters are still [initialized in an nn.Linear layer and copied over](https://github.com/microsoft/tutel/pull/210/files#diff-ef4f1a9825a158f9757d26b3d77b010752b142e479e9167c55b501829056d93aR40-R47)

### CODE REVIEW

**Comment by vchiley** [comment]
[Here](https://github.com/microsoft/tutel/pull/210#issuecomment-1742179844) @ghostplant claims: 
> Original's expert parameter initialization seed was specially designed to ensure initial parameter values to be deterministic regardless of whether expert parameters are sharded or not. In order words, 2 GPU training 2 experts in total should have exact the same training loss compared with 4 GPU training 2 experts (i.e. each GPU training half of an expert).

I don't see how original init ever had this functionality.

Regardless, this change does not change how seeding happens. If the functionality previously existed, it still will.

The functionality @ghostplant claims is possible if every GPU had the same seed, then in `reset_parameters` we instantiate a tensor with ALL experts, and grab ONLY the subset required for the current GPU.

File: tutel/experts/ffn.py [hunk 1]
```diff
@@ -18,34 +23,32 @@ def __init__(self, hidden_size_per_expert, activation_fn=None, activation_fn_wit
...
+                if self.batched_fc1_bias is not None:
+                    self.batched_fc1_bias[i] = fc1.bias
+                self.batched_fc2_w[i] = fc2.weight.t()
+                if self.batched_fc2_bias is not None:
+                    self.batched_fc2_bias[i] = fc2.bias[:self.batched_fc2_bias.size(-1)]
 
     def extra_repr(self):
-        return 'model_dim=%d, hidden_size=%d, output_dim=%d, local_experts=%d' % (
-            self.batched_fc1_w.size(2), self.batched_fc1_w.size(1), self.batched_fc2_w.size(2), self.batched_fc1_w.size(0)
+        return 'model_dim=%d, hidden_size=%d, output_dim=%d, local_experts=%d. has bias=%s.' % (
```
Comment (ghostplant) [inline tutel/experts/ffn.py:50]: What about has_bias (with underline) instead?

File: tutel/impls/moe_layer.py [hunk 2]
```diff
@@ -167,10 +167,11 @@ def __init__(
                 assert 'fused_custom_fn' not in experts, "`fused_custom_fn` option for Tutel Moe-layer has been deprecated, please follows helloworld_from_scratch.py for custom construction instead."
                 assert 'implicit_dropout_p' not in experts, "`implicit_dropout_p` option for Tutel Moe-layer has been deprecated, please use torch.nn.Dropout(p=implicit_dropout_p) on custom activation_fn (for fc1_dropout) and after Tutel Moe-layer (for fc2_dropout) instead."
 
+            experts['model_dim'] = self.model_dim
+            experts['local_experts'] = self.num_local_experts
+            experts['sharded_count'] = self.sharded_count
```
Comment (ghostplant) [inline tutel/impls/moe_layer.py:172]: How to explicitly set `bias=False`?

File: tutel/experts/ffn.py [hunk 3]
```diff
@@ -18,34 +23,32 @@ def __init__(self, hidden_size_per_expert, activation_fn=None, activation_fn_wit
...
-        self.register_parameter(name='batched_fc2_bias', param=torch.nn.Parameter(fc2_bias.squeeze(0)))
+        self.batched_fc1_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, model_dim))
+        self.batched_fc2_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, self.output_dim))
+        if bias:
+            self.batched_fc1_bias = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size))
+            self.batched_fc2_bias = torch.nn.Parameter(torch.empty(local_experts, (self.output_dim + sharded_count - 1) // sharded_count))
+        else:
+            self.register_parameter('batched_fc1_bias', None)
+            self.register_parameter('batched_fc2_bias', None)
+
```
Comment (ghostplant) [inline tutel/experts/ffn.py:34]: Seems like there are 2 options, right?
(1) the same as orignial, having both fc_bias_1 and fc_bias_2;
(2) none of fc_bias_1 and fc_bias_2 is included in the network;

Comment (vchiley) [reply tutel/experts/ffn.py:34]: correct. I feel like adding 2 bias flags is extra.
if people want one of the biases but not the other, they can instantiate the network with both biases (`bias=True`) then remove one eg something like `model.layer.moe.experts.register_parameter('batched_fc2_bias', None)`

File: tutel/experts/ffn.py (hunk 1 above)
Comment (vchiley) [inline tutel/experts/ffn.py:50]:
```suggestion
        return 'model_dim=%d, hidden_size=%d, output_dim=%d, local_experts=%d. has_bias=%s.' % (
```

File: tutel/impls/moe_layer.py (hunk 2 above)
Comment (vchiley) [reply tutel/impls/moe_layer.py:172]: `experts` arg passed into the `MOELayer.__init__` is treated as a  dict of kwargs which will be `**` expanded into the expert init below: `self.experts = fused_experts.ExpertModule(**experts)`.

This kwarg dict should set bias (if you want to set bias).

```
            experts['model_dim'] = self.model_dim
            experts['local_experts'] = self.num_local_experts
            experts['sharded_count'] = self.sharded_count
```
are set from within the context layer because they are defined in the MOELayer context therefore this is the correct place to set these.

File: tutel/experts/ffn.py [hunk 4]
```diff
@@ -18,34 +23,32 @@ def __init__(self, hidden_size_per_expert, activation_fn=None, activation_fn_wit
...
+                    self.batched_fc1_bias[i] = fc1.bias
+                self.batched_fc2_w[i] = fc2.weight.t()
+                if self.batched_fc2_bias is not None:
+                    self.batched_fc2_bias[i] = fc2.bias[:self.batched_fc2_bias.size(-1)]
 
     def extra_repr(self):
-        return 'model_dim=%d, hidden_size=%d, output_dim=%d, local_experts=%d' % (
-            self.batched_fc1_w.size(2), self.batched_fc1_w.size(1), self.batched_fc2_w.size(2), self.batched_fc1_w.size(0)
+        return 'model_dim=%d, hidden_size=%d, output_dim=%d, local_experts=%d. has_bias=%s.' % (
+            self.batched_fc1_w.size(2), self.batched_fc1_w.size(1), self.batched_fc2_w.size(2), self.batched_fc1_w.size(0), self.batched_fc1_bias is not None
```
Comment (ghostplant) [inline tutel/experts/ffn.py:51]: Please turn to 2 items: `has_fc1_bias` and `has_fc2_bias`

File: tutel/experts/ffn.py (hunk 3 above)
Comment (ghostplant) [reply tutel/experts/ffn.py:34]: Please turn `bias=True` to two cases, and update `if .. else` as well.

File: tutel/impls/moe_layer.py (hunk 2 above)
Comment (vchiley) [reply tutel/impls/moe_layer.py:172]: `has_fc1_bias` and `has_fc2_bias` both default to True.

File: tutel/experts/ffn.py (hunk 3 above)
Comment (vchiley) [reply tutel/experts/ffn.py:36]: [updated arg: `bias` to `has_fc1_bias` and `has_fc2_bias`](https://github.com/microsoft/tutel/pull/216/commits/9a3686e9ca3a8d3d5f729c81d1090f756f89d497) (both default to `True`).

**Comment by ghostplant** [comment]
Thanks! Please wait for checkpoint compatibility validation.

File: tutel/experts/ffn.py [hunk 5]
```diff
@@ -18,34 +23,35 @@ def __init__(self, hidden_size_per_expert, activation_fn=None, activation_fn_wit
...
-        self.register_parameter(name='batched_fc2_w', param=torch.nn.Parameter(fc2_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc1_bias', param=torch.nn.Parameter(fc1_bias.squeeze(0)))
-        self.register_parameter(name='batched_fc2_bias', param=torch.nn.Parameter(fc2_bias.squeeze(0)))
+        self.batched_fc1_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, model_dim))
+        self.batched_fc2_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, self.output_dim))
+        if has_fc1_bias:
+             self.batched_fc1_bias = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size))
+        else:
+             self.register_parameter('batched_fc1_bias', None)
+        if has_fc2_bias:
```
Comment (ghostplant) [inline tutel/experts/ffn.py:32]: Can you remove this ending space at line 32?

File: tutel/experts/ffn.py [hunk 6]
```diff
@@ -18,34 +23,35 @@ def __init__(self, hidden_size_per_expert, activation_fn=None, activation_fn_wit
...
-            fc2_weight[0, i, :, :], fc2_bias[0, i, :] = fc2.weight.t(), fc2.bias[:fc2_bias.size(-1)]
-
-        self.register_parameter(name='batched_fc1_w', param=torch.nn.Parameter(fc1_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc2_w', param=torch.nn.Parameter(fc2_weight.squeeze(0)))
-        self.register_parameter(name='batched_fc1_bias', param=torch.nn.Parameter(fc1_bias.squeeze(0)))
-        self.register_parameter(name='batched_fc2_bias', param=torch.nn.Parameter(fc2_bias.squeeze(0)))
+        self.batched_fc1_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, model_dim))
+        self.batched_fc2_w = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size, self.output_dim))
+        if has_fc1_bias:
+             self.batched_fc1_bias = torch.nn.Parameter(torch.empty(local_experts, self.hidden_size))
```
Comment (ghostplant) [inline tutel/experts/ffn.py:29]: What is the reason not using `self.register_parameter('batched_fc1_bias', param=torch.nn.Parameter(torch.empty(local_experts, self.hidden_size)))`?

**Comment by ghostplant** [comment]
Please apply or accept this change: https://github.com/vchiley/tutel_msr/pull/2. Then we can merge it. Thanks!

File: tutel/experts/ffn.py (hunk 6 above)
Comment (vchiley) [reply tutel/experts/ffn.py:29]: param instantiation follows PyTorch convention eg [nn.Linear](https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/linear.py#L96)

Let me know if you want to break from PyTorch convention

File: tutel/experts/ffn.py (hunk 5 above)
Comment (vchiley) [inline tutel/experts/ffn.py:32]:
```suggestion
        if has_fc2_bias:
```

File: tutel/experts/ffn.py (hunk 6 above)
Comment (ghostplant) [reply tutel/experts/ffn.py:29]: OK, seems like there is tiny difference when paramter definitions are outside __init__ function (https://discuss.pytorch.org/t/nn-module-add-new-parameter-setattr-vs-register-parameter/108238). But inside __init__ function, they seem to be the same result set.

**Comment by ghostplant** [comment]
Thanks for the contribution!
